### ISSUE TITLE
Provide a common abstraction of TripleSource and QuadSource

### ISSUE THREAD

**Issue by MattesWhite** [pchampin/sophia_rs#52]
This PR originates from the discussion in #49 starting with [this comment](https://github.com/pchampin/sophia_rs/pull/49#issuecomment-595701358).

# Summary

- It would be beneficial to have convenience adapters for general tasks on `*Sources`, e.g.:
  - rename bnodes into fresh ones
  - replace bnodes with variables / variables with bnodes
  - change absolute IRIs to relative ones (given a base IRI) (opposite of resolve)
  - replace variables by their bound values (given a mapping)
  - resolve IRIs given a base IRI
- The current crate structure required to duplicate much of the code for adapters (and `streaming_mode`).

# Problems

1. **Implementing traits with associated traits does not consider trait bounds.** Meaning is there is an adapter `Normalizer<S>` one can only implement either `Iterator<Item = Triple> for Normalizer<S: TripleSource>` or `Iterator<Item = Quad> for Normalizer<S: QuadSource>`. Both together would result in `conflicting implementation`.
1. **Splitted implementations violate privacy bounds.** Declaring a `struct` in `crate::triple::stream` and implementing `QuadSource` in `crate::quad::stream` requires the fields to be `pub(crate)` at least.

# Solutions

**For 1.:** Maybe some overall abstraction for statements?

**For 2.:** Merge the `stream` submodules into a top-level module `stream` equal to `triple` and `quad`.

**Comment by pchampin** [comment]
> Merge the stream submodules into a top-level module stream equal to triple and quad.

I started trying this out, factorizing some of the code into a `StatementSource` trait, then defining specializations of it as `TripleSource` and `QuadSource`. But I ran into the following problem:

One can not easily use the specialized traits in a trait bound. Instead of writing
```rust
        pub fn consume_triple_source<TS: TripleSource>(mut ts: TS) 
        { /*...*/ }
````
one needs to write
```rust
        pub fn consume_triple_source<TS: TripleSource>(mut ts: TS) 
        where <TS::Mode as StreamingMode>::Item: UnsafeTriple
        { /*...*/ }
````
which is unintuitive and cumbersome. This does not go in the direction of making the API easier...

I put [a summarized version of this problem](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=cb7ee5fdcb52bfff47d8f143ac7fd461) in the playground.
